### PR TITLE
Basic implementation of banners, including API to modify them with ease.

### PR DESCRIPTION
**Pull request by Sandertv**
## Introduction

This pull request aims to add a basic banner implementation with API so developers can easily modify banners.

### API changes

Registers and adds items:
- Banner

Registers and adds blocks:
- StandingBanner;
- WallBanner (extends StandingBanner)

Registers and adds tiles:
- Banner

The banner tile and item contain documented API, and constants for banner patterns and colours in the banner tile. These colours differ from normal (for example) wool colours. The documentation on these functions should speak for themselves.

### Behavioural changes

Banners can now be placed and modified through plugins, crafting of those **does not work yet**.

## Follow-up

All functions have been tested to work as intended, no more tests should be required.

## Tests

<details><summary>[spoiler] edited by @dktapps to not have to scroll 17 miles to the comments</summary>

```
$tile->changePattern($this->patternId, Banner::PATTERN_BORDER, 7);

Pre:
  ["Base"]=>
  object(pocketmine\nbt\tag\IntTag)#19800 (2) {
    ["__name":protected]=>
    string(4) "Base"
    ["value":protected]=>
    int(12)
  }
Post:
["Base"]=>
  object(pocketmine\nbt\tag\IntTag)#19800 (2) {
    ["__name":protected]=>
    string(4) "Base"
    ["value":protected]=>
    int(7)
  }
```
```
$this->patternId = $tile->addPattern(Banner::PATTERN_GRADIENT, mt_rand(0, 15));

Pre:
["Patterns"]=>
  object(pocketmine\nbt\tag\ListTag)#20254 (3) {
    ["tagType":"pocketmine\nbt\tag\ListTag":private]=>
    int(0)
    ["__name":protected]=>
    string(8) "Patterns"
    ["value":protected]=>
    NULL
  }
Post:
["Patterns"]=>
  object(pocketmine\nbt\tag\ListTag)#20254 (4) {
    ["tagType":"pocketmine\nbt\tag\ListTag":private]=>
    int(10)
    ["__name":protected]=>
    string(8) "Patterns"
    ["value":protected]=>
    NULL
    ["0"]=>
    object(pocketmine\nbt\tag\CompoundTag)#20067 (4) {
      ["__name":protected]=>
      string(0) ""
      ["value":protected]=>
      NULL
      ["Color"]=>
      object(pocketmine\nbt\tag\IntTag)#20120 (2) {
        ["__name":protected]=>
        string(5) "Color"
        ["value":protected]=>
        int(10)
      }
      ["Pattern"]=>
      object(pocketmine\nbt\tag\StringTag)#20070 (2) {
        ["__name":protected]=>
        string(7) "Pattern"
        ["value":protected]=>
        string(3) "gra"
      }
    }
  }
```
```
$tile->changePattern($this->patternId, Banner::PATTERN_BORDER, 7);

Pre:
["Patterns"]=>
  object(pocketmine\nbt\tag\ListTag)#20254 (4) {
    ["tagType":"pocketmine\nbt\tag\ListTag":private]=>
    int(10)
    ["__name":protected]=>
    string(8) "Patterns"
    ["value":protected]=>
    NULL
    ["0"]=>
    object(pocketmine\nbt\tag\CompoundTag)#20067 (4) {
      ["__name":protected]=>
      string(0) ""
      ["value":protected]=>
      NULL
      ["Color"]=>
      object(pocketmine\nbt\tag\IntTag)#20120 (2) {
        ["__name":protected]=>
        string(5) "Color"
        ["value":protected]=>
        int(10)
      }
      ["Pattern"]=>
      object(pocketmine\nbt\tag\StringTag)#20070 (2) {
        ["__name":protected]=>
        string(7) "Pattern"
        ["value":protected]=>
        string(3) "gra"
      }
    }
  }
Post:
["Patterns"]=>
  object(pocketmine\nbt\tag\ListTag)#20254 (4) {
    ["tagType":"pocketmine\nbt\tag\ListTag":private]=>
    int(10)
    ["__name":protected]=>
    string(8) "Patterns"
    ["value":protected]=>
    NULL
    ["0"]=>
    object(pocketmine\nbt\tag\CompoundTag)#20067 (4) {
      ["__name":protected]=>
      string(0) ""
      ["value":protected]=>
      NULL
      ["Color"]=>
      object(pocketmine\nbt\tag\IntTag)#20133 (2) {
        ["__name":protected]=>
        string(5) "Color"
        ["value":protected]=>
        int(7)
      }
      ["Pattern"]=>
      object(pocketmine\nbt\tag\StringTag)#19315 (2) {
        ["__name":protected]=>
        string(7) "Pattern"
        ["value":protected]=>
        string(2) "bo"
      }
    }
  }
```
```
$tile->deleteBottomPattern();

Pre:
["Patterns"]=>
  object(pocketmine\nbt\tag\ListTag)#20254 (4) {
    ["tagType":"pocketmine\nbt\tag\ListTag":private]=>
    int(10)
    ["__name":protected]=>
    string(8) "Patterns"
    ["value":protected]=>
    NULL
    ["0"]=>
    object(pocketmine\nbt\tag\CompoundTag)#20067 (4) {
      ["__name":protected]=>
      string(0) ""
      ["value":protected]=>
      NULL
      ["Color"]=>
      object(pocketmine\nbt\tag\IntTag)#20133 (2) {
        ["__name":protected]=>
        string(5) "Color"
        ["value":protected]=>
        int(7)
      }
      ["Pattern"]=>
      object(pocketmine\nbt\tag\StringTag)#19315 (2) {
        ["__name":protected]=>
        string(7) "Pattern"
        ["value":protected]=>
        string(2) "bo"
      }
    }
  }
Post:
["Patterns"]=>
  object(pocketmine\nbt\tag\ListTag)#20254 (3) {
    ["tagType":"pocketmine\nbt\tag\ListTag":private]=>
    int(10)
    ["__name":protected]=>
    string(8) "Patterns"
    ["value":protected]=>
    NULL
  }
```
</details>
<br>
Results were gathered using this test plugin: https://github.com/BlockHorizons/BannerTests

### CODE REVIEW

File: src/pocketmine/block/StandingBanner.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	public function getVariantBitmask() : int{
+		return 0;
+	}
+
+	public function getDrops(Item $item) : array{
+		return [];
+	}
+
+	public function onBreak(Item $item, Player $player = null) : bool{
+		$this->level->dropItem($this, ItemFactory::get(Item::BANNER)->setNamedTag($this->level->getTile($this)->namedtag));
```
Comment (dktapps) [inline src/pocketmine/block/StandingBanner.php]: don't be too sure there's a tile to read from... if a plugin uses setBlock() this will cause problems.

Comment (Sandertv) [reply src/pocketmine/block/StandingBanner.php]: Ah right, I'll add a check for that. Thanks.

File: src/pocketmine/block/WallBanner.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+	protected $id = self::WALL_BANNER;
+
+	public function getName() : string{
+		return "Wall Banner";
+	}
+
+	public function onUpdate(int $type){
+		$faces = [
+			2 => 3,
```
Comment (jasonwynn10) [inline src/pocketmine/block/WallBanner.php]: Could probably use Vector3 constants here

File: src/pocketmine/block/StandingBanner.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+		return 0;
+	}
+
+	public function getDrops(Item $item) : array{
+		return [];
+	}
+
+	public function onBreak(Item $item, Player $player = null) : bool{
+		if(($tile = $this->level->getTile($this)) !== null) {
+			$this->level->dropItem($this, ItemFactory::get(Item::BANNER)->setNamedTag($tile->namedtag));
```
Comment (dktapps) [inline src/pocketmine/block/StandingBanner.php]: This will still retain coordinate NBT.

**Review by jasonw4331 — CHANGES_REQUESTED** [review]

File: src/pocketmine/block/StandingBanner.php (hunk 3 above)
Comment (Sandertv) [reply src/pocketmine/block/StandingBanner.php]: Ah, that's true.

**Comment by matcracker** [comment]
Good job! 👍

File: src/pocketmine/tile/Banner.php [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	/**
+	 * Deletes the bottom pattern of the banner.
+	 *
+	 * @return bool indicating whether the banner was empty or not.
+	 */
+	public function deleteBottomPattern() : bool{
+		if(empty((array) $this->namedtag->Patterns)){
+			return false;
+		}
+		$index = (int) min(array_keys((array) $this->namedtag->Patterns));
```
Comment (Muqsit) [inline src/pocketmine/tile/Banner.php]: Just gone through the code, I don't think the (int) casting is needed.

Comment (Sandertv) [reply src/pocketmine/tile/Banner.php]: Sometimes I would get "\0" here, which would lead to an error. I'm not entirely sure of the reason, but this fixed it.

Comment (SOF3) [reply src/pocketmine/tile/Banner.php]: Note that all strings starting with non-numeric bytes, including non-printable bytes ("\0" and "\1"), are converted to int(0) when cast to int. Make sure you don't make everything zero.

**Comment by Sandertv** [comment]
Sadly due to a texture bug with banners in Minecraft PE, the actual result of deleting patterns and changing them is not properly visible, and applying multiple same typed patterns on top of eachother turns the banner invisible with multiple texture bug. This does also affect vanilla, so it's nothing I can do about. I'll get some testing results to show that it does work on our end.

**Comment by Sandertv** [comment]
screenshot_20170828-203457

**Comment by Sandertv** [comment]
Added some tests showing that the code works well. Sadly the textures of Minecraft PE don't. (yet)

File: src/pocketmine/tile/Banner.php (hunk 4 above)
Comment (Sandertv) [reply src/pocketmine/tile/Banner.php]: Properly fixed now.

File: src/pocketmine/block/WallBanner.php (hunk 2 above)
Comment (Sandertv) [reply src/pocketmine/block/WallBanner.php]: The PillarRotationHelper only works for pillars; It takes 3 rotations in account, facing north, facing east or facing upwards. That doesn't really work with banners that have 4 rotations; north, east, south and west.

**Comment by Sandertv** [comment]
As far as I can tell this is ready to be merged, unless anybody wants me to change or add anything of course.

File: src/pocketmine/block/WallBanner.php (hunk 2 above)
Comment (jasonwynn10) [reply src/pocketmine/block/WallBanner.php]: what about vector3 constants? https://github.com/pmmp/PocketMine-MP/blob/master/src/pocketmine/math/Vector3.php#L30L33

Comment (Sandertv) [reply src/pocketmine/block/WallBanner.php]: For the sake of consistency I think I'm just going to keep it like this; The WallSign class has the exact same piece of code.

Comment (jasonwynn10) [reply src/pocketmine/block/WallBanner.php]: They probably won't be there for long 😛 
https://github.com/pmmp/PocketMine-MP/commit/c394aea80325d9cf06aab4cee302e4be5a82e56b

**Comment by Sandertv** [comment]
Final update, since the last minecraft Beta lots of banner bugs got fixed, deleting and changing patterns works flawlessly now, only adding a lot of the same patterns on top of eachother with the same patterns sometimes causes some visual glitches, but that's not in my hands.

File: src/pocketmine/tile/Tile.php [hunk 5]
```diff
@@ -76,6 +77,7 @@ public static function init(){
 		self::registerTile(ItemFrame::class);
 		self::registerTile(Sign::class);
 		self::registerTile(Skull::class);
+		self::registerTile(Banner::class);
```
Comment (dktapps) [inline src/pocketmine/tile/Tile.php]: alphabetical order please.

**Review by jasonw4331 — APPROVED** [review]

File: src/pocketmine/block/StandingBanner.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	public function place(Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $facePos, Player $player = null) : bool{
+		if($face !== Vector3::SIDE_DOWN){
+			$nbt = new CompoundTag("", [
+				new StringTag("id", Tile::BANNER),
+				new IntTag("x", $blockReplace->x),
+				new IntTag("y", $blockReplace->y),
+				new IntTag("z", $blockReplace->z),
+				isset($item->getNamedTag()->Base) ? $item->getNamedTag()->Base : new IntTag("Base", $item->getDamage() & 0x0f),
```
Comment (TheDiamondYT1) [inline src/pocketmine/block/StandingBanner.php]: Couldn't you use null coalescing here?

**Review by TheDiamondYT1 — APPROVED** [review]

File: src/pocketmine/tile/Banner.php [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+	/**
+	 * Returns an array containing all pattern IDs
+	 *
+	 * @return array
+	 */
+	public function getPatternIds() : array{
+		$keys = array_keys((array) $this->namedtag->Patterns);
+
+		foreach($keys as $key => $index){
```
Comment (dktapps) [inline src/pocketmine/tile/Banner.php]: array_filter() could be used here.

**Comment by dktapps** [comment]
Resolved merge conflicts, please pull before making any further changes.

**Comment by dktapps** [comment]
closed and reopened to (attempt to) kick travis into gear

**Comment by dktapps** [comment]
well, that worked... not

File: src/pocketmine/block/StandingBanner.php [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	public function getVariantBitmask() : int{
+		return 0;
+	}
+
+	public function getDrops(Item $item) : array{
+		return [];
+	}
+
+	public function onBreak(Item $item, Player $player = null) : bool{
+		if(($tile = $this->level->getTile($this)) !== null) {
```
Comment (dktapps) [inline src/pocketmine/block/StandingBanner.php]: nit: formatting issue

File: src/pocketmine/block/WallBanner.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+		return "Wall Banner";
+	}
+
+	public function onUpdate(int $type){
+		$faces = [
+			Vector3::SIDE_NORTH => 3,
+			Vector3::SIDE_SOUTH => 2,
+			Vector3::SIDE_WEST => 5,
+			Vector3::SIDE_EAST => 4
+		];
```
Comment (dktapps) [inline src/pocketmine/block/WallBanner.php]: This can be done with a simple XOR bit operation instead of needing a map (`this->meta ^ 0x01`).

File: src/pocketmine/item/Banner.php [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+use pocketmine\block\Block;
+use pocketmine\block\BlockFactory;
+use pocketmine\nbt\tag\CompoundTag;
+use pocketmine\nbt\tag\IntTag;
+use pocketmine\nbt\tag\ListTag;
+use pocketmine\nbt\tag\StringTag;
+
+class Banner extends Item{
+
+	public function __construct(int $meta = 0) {
```
Comment (dktapps) [inline src/pocketmine/item/Banner.php]: nit: formatting issue

File: src/pocketmine/tile/Tile.php [hunk 11]
```diff
@@ -170,9 +172,8 @@ public function getCleanedNBT() : ?CompoundTag{
 		$tag->removeTag("x", "y", "z", "id");
 		if($tag->getCount() > 0){
 			return $tag;
-		}else{
-			return null;
 		}
+		return null;
```
Comment (dktapps) [inline src/pocketmine/tile/Tile.php]: this change doesn't look related to the PR?

File: src/pocketmine/item/Banner.php [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+	 * Returns the total count of patterns on this banner.
+	 *
+	 * @return int
+	 */
+	public function getPatternCount() : int{
+		return count($this->getPatternIds());
+	}
+
+	public function correctNBT() : void{
+		$tag = $this->getNamedTag() ?? new CompoundTag();
```
Comment (dktapps) [inline src/pocketmine/item/Banner.php]: this is no longer necessary

File: src/pocketmine/item/Banner.php [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+		if(!isset($tag->Base) or !($tag->Base instanceof IntTag)) {
```
Comment (dktapps) [inline src/pocketmine/item/Banner.php]: nit: formatting issue

File: src/pocketmine/item/Banner.php [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+			$tag->Base = new IntTag("Base", $this->meta);
+		}
+
+		if(!isset($tag->Patterns) or !($tag->Patterns instanceof ListTag)) {
```
Comment (dktapps) [inline src/pocketmine/item/Banner.php]: and another here

File: src/pocketmine/block/StandingBanner.php [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+	public function place(Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $facePos, Player $player = null) : bool{
+		if($face !== Vector3::SIDE_DOWN){
+			$nbt = new CompoundTag("", [
+				new StringTag("id", Tile::BANNER),
+				new IntTag("x", $blockReplace->x),
+				new IntTag("y", $blockReplace->y),
+				new IntTag("z", $blockReplace->z),
+				$item->getNamedTag()->Base ?? new IntTag("Base", $item->getDamage() & 0x0f),
+			]);
```
Comment (dktapps) [inline src/pocketmine/block/StandingBanner.php]: This should make use of the new API for creating tile NBT.

**Review by dktapps — DISMISSED** [review]
Please see my comments on the diff.

File: src/pocketmine/item/Banner.php [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+	 *
+	 * @return array
+	 */
+	public function getPatternData(int $patternId) : array{
+		if(!$this->patternExists($patternId)){
+			return [];
+		}
+
+		return [
+			self::TAG_COLOR => $this->getNamedTag()->Patterns->{$patternId}->getInt(self::TAG_COLOR, 0),
```
Comment (dktapps) [inline src/pocketmine/item/Banner.php]: why does this have a default and the one below doesn't?

**Comment by dktapps** [comment]
Context: Apparently Travis CI has flagged this PR as abusung their workers, so Travis is refusing to build the most recent commits.

**Review by dktapps — APPROVED** [review]
